llvm/lib/Transforms/Instrumentation/AddressSanitizer.cpp

```diff
@@ -1234,7 +1245,91 @@ static bool isUnsupportedAMDGPUAddrspace(Value *Addr) {
   return false;
 }
 
+static bool isUnsupportedSPIRAddrspace(Value *Addr, Function *Func) {
+  Type *PtrTy = cast<PointerType>(Addr->getType()->getScalarType());
+  if (PtrTy->getPointerAddressSpace() != 0) {
+    // Skip SPIR-V built-in varibles
+    StringRef Name;
+    if (auto *CE = dyn_cast<ConstantExpr>(Addr)) {
+      const auto OpCode = CE->getOpcode();
+      if (OpCode == Instruction::GetElementPtr ||
+          OpCode == Instruction::AddrSpaceCast) {
+        Name = CE->getOperand(0)->getName();
+      }
+    } else {
+      Name = Addr->getName();
+    }
+    if (Name.starts_with("__spirv_BuiltIn")) {
+      return true;
+    }
+  } else {
+    // FIXME: Check kernel arguments instead of value's name
+    auto Name = Addr->getName();
+    if (Name == "_arg_NumWorkItems" || Name == "_arg_KernelFunc") {
+      return true;
+    }
+  }
+  return false;
+}
+
+GlobalVariable *AddressSanitizer::GetOrCreateGlobalString(Module &M, StringRef Name,
+                                               StringRef Value,
+                                               unsigned AddressSpace) {
+  GlobalVariable *StringGV = nullptr;
+  if (GlobalStringMap.find(Value.str()) != GlobalStringMap.end()) {
+    StringGV = GlobalStringMap[Value.str()];
+  } else {
+    auto *Ty =
+        ArrayType::get(Type::getInt8Ty(M.getContext()), Value.size() + 1);
+    StringGV = new GlobalVariable(
+        M, Ty, true, GlobalValue::InternalLinkage,
+        ConstantDataArray::getString(M.getContext(), Value), Name, nullptr,
+        GlobalValue::NotThreadLocal, AddressSpace);
+    GlobalStringMap[Value.str()] = StringGV;
+  }
+  return StringGV;
+}
+
+void AddressSanitizer::AppendDebugInfoToArgs(Instruction *InsertBefore, Value *Addr,
+                                  SmallVector<Value *> &Args) {
+  auto *M = InsertBefore->getModule();
+  auto &C = InsertBefore->getContext();
+  auto &Loc = InsertBefore->getDebugLoc();
+  constexpr unsigned ConstantAS = 2;
+  auto *I8PtrTy = Type::getInt8Ty(C)->getPointerTo(ConstantAS);
+
+  // Address Space
+  Type *PtrTy = cast<PointerType>(Addr->getType()->getScalarType());
+  Args.push_back(
+      ConstantInt::get(Type::getInt32Ty(C), PtrTy->getPointerAddressSpace()));
+
+  // File & Line
+  if (Loc) {
+    auto FileName = Loc->getFilename();
```

Comment:
[Don't almost always use `auto`](https://llvm.org/docs/CodingStandards.html#use-auto-type-deduction-to-make-code-more-readable), the type of `FileName` is not immediately obvious here.

I haven't even been able to figure it out. `DebugLoc::operator->` is overloaded, so it returns `DILocation`, but even there I wasn't able to find `getFilename` method. But perhaps I've just missed something.

Reply:
I don't understand what you mean. I can find `getFilename` method in `DILocation`, and since I use clangd with VSCode, I know here the type of `FileName` is `StringRef`.

Reply:
> since I use `clangd` with `VSCode`, I know here the type of `FileName` is `StringRef`.

If you have to use a tool for that, then the type is not immediately obvious. Not everyone uses `VSCode`, not everyone uses `clangd`, a lot of code review happens here on GitHub directly, etc. I would define immediately obvious as "can be inferred by looking into this and surrounding lines of code in plain text".

> I can find `getFilename` method in `DILocation`

I used [`DILocation` doxygen](https://llvm.org/doxygen/classllvm_1_1DILocation.html) page to check for the method and apparently the page is buggy: the method is indeed in the codebase, but by some reason it is not present on the page.